src/nvim/buffer.c

```diff
@@ -309,7 +309,7 @@ close_buffer (
   // buffers.
   // The caller must take care of NOT deleting/freeing when 'bufhidden' is
   // "hide" (otherwise we could never free or delete a buffer).
-  if (!buf->terminal) {
+  if (!buf->terminal && !buf->b_messages) {
```

Comment:
I was actually thinking about having one variable like `buf->b_internal_type` which is an int for un-editable buffers.  `BUF_TERMINAL`, `BUF_MESSAGES`, etc.  If it was non-zero then the same rules would apply to all of them.